crates/client/block_import/src/pre_validate.rs

```diff
@@ -0,0 +1,336 @@
+
+    if let Some(expected) = block.commitments.event_count {
+        if expected != events_with_tx_hash.len() as u64 {
+            return Err(BlockImportError::EventCount { got: events_with_tx_hash.len() as _, expected });
+        }
+    }
+
+    let got = if events_with_tx_hash.is_empty() {
+        Felt::ZERO
+    } else if block.header.protocol_version < StarknetVersion::STARKNET_VERSION_0_13_2 {
+        let events_hash =
+            events_with_tx_hash.into_par_iter().map(|(_, event)| event.compute_hash_pedersen()).collect::<Vec<_>>();
+        compute_merkle_root::<Pedersen>(&events_hash)
+    } else {
+        let events_hash = events_with_tx_hash
+            .into_par_iter()
+            .map(|(hash, event)| event.compute_hash_poseidon(&hash))
+            .collect::<Vec<_>>();
+        compute_merkle_root::<Poseidon>(&events_hash)
+    };
+
+    if let Some(expected) = block.commitments.event_commitment {
+        if expected != got {
+            return Err(BlockImportError::EventCommitment { got, expected });
+        }
+    }
+
+    Ok(got)
+}
+
+/// Compute the receipt commitment for a block.
+fn receipt_commitment(block: &UnverifiedFullBlock, _validation: &Validation) -> Result<Felt, BlockImportError> {
+    let hashes = block.receipts.par_iter().map(TransactionReceipt::compute_hash).collect::<Vec<_>>();
+    let got = compute_merkle_root::<Poseidon>(&hashes);
+
+    if let Some(expected) = block.commitments.receipt_commitment {
+        if expected != got {
+            return Err(BlockImportError::ReceiptCommitment { got, expected });
+        }
+    }
+    Ok(got)
+}
+
+/// Compute the state diff commitment for a block.
+fn state_diff_commitment(block: &UnverifiedFullBlock, _validation: &Validation) -> Result<Felt, BlockImportError> {
+    let got = block.state_diff.len() as u64;
+    if let Some(expected) = block.commitments.state_diff_length {
+        if expected != got {
+            return Err(BlockImportError::StateDiffLength { got, expected });
+        }
+    }
+
+    let got = block.state_diff.compute_hash();
+    if let Some(expected) = block.commitments.state_diff_commitment {
+        if expected != got {
+            return Err(BlockImportError::StateDiffCommitment { got, expected });
+        }
+    }
+    Ok(got)
+}
+
+/// Compute the root hash of a list of values.
+// The `HashMapDb` can't fail, so we can safely unwrap the results.
+//
+// perf: Note that committing changes still has the greatest performance hit
+// as this is where the root hash is calculated. Due to the Merkle structure
+// of Bonsai Tries, this results in a trie size that grows very rapidly with
+// each new insertion. It seems that the only vector of optimization here
+// would be to parallelize the tree traversal on insertion and optimize hash computation.
+// It seems lambdaclass' crypto lib does not do simd hashing, we may want to look into that.
+fn compute_merkle_root<H: StarkHash + Send + Sync>(values: &[Felt]) -> Felt {
```

Comment:
I think not, because i don't see any other crate than block import using it

in fact the i want to make bonsai-trie an implementation detail that only block import and db care about

Reply:
dc-block-import could be considered part of the madara backend along with dc-db

Reply:
using very loose quotation marks

Reply:
ah i thought you were talking about the global tries

Reply:
agreed that it should probably live in its own crate and also with more unit tests

Reply:
i don't know if adding unit tests for that would be *that* useful this is the very basic usage of bonsai-trie that is already tested inside the bonsai-trie crate

i don't know if making a crate is a good idea, i would actually prefer if we lowered the number of primitive crates we currently have
and also i don't see why any other crate would ever want to compute a transaction commitment except the crate creating and validating blocks